### DISCUSSION TITLE
How can I deal with this nonsensical errors that I'm constantly getting?

### DISCUSSION THREAD

**Discussion by john01dav** [parcel-bundler/parcel#5385] [Q&A]
As I use Parcel, I'm getting a _constant_ stream of nonsensical errors. Sometimes it gives bogus output, and sometimes it just flat out refuses to build with no useful feedback.

Starting yesterday, I found that Bootstrap CSS quit loading in one HTML file, but it worked in another with _identical contents_ (I have each HTML file set as an entry point), until it randomly stopped working in the other files with just a rebuild and no code changes. Eventually, somehow, after many hours upon hours of painstaking testing I managed to figure out that having "babel-polyfill" in the dependencies causes that, but removing that causes runtime errors in another file that, when I look them up in a search engine, says to add "babel-polyfill." This should _never_ happen in a production-grade project, so perhaps one may think that I should downgrade to v1, but I can't do that because of the deasync bug that isn't fixed! It _literally_ crashes on my system, but of course only sometimes because everyone just _loves_ a Schrödinger's bug.

Just now, I'm getting this error:
```
🚨 Build failed.
@parcel/bundler-default: The expression evaluated to a falsy value:

  (0, _assert().default)(context != null)

AssertionError [ERR_ASSERTION]: The expression evaluated to a falsy value:

  (0, _assert().default)(context != null)

```
I have no clue what this is saying about my code, nor can I find any relevant issue report about it, so I'm basically stuck trying literally random things until hopefully it eventually works, until the next thing breaks. 

Before that, I was getting errors about not finding nodes with what appears to be either hashes or UUIDs of those nodes. Again, who knows what this actually means, so I can't debug it. Eventually, after trying random things, I found that I needed to update to the latest Parcel nightly to make it go away, but if I downgrade again the error doesn't return (until it does, a bit later, seemingly randomly).

As such, I ask: how can I use Parcel when I _constantly_ get these errors? When I fix one it's never more than an hour until the next one pops up, which then takes *at least* a day to fix. I can't be productive like this, but there is no alternative to Parcel, it seems (Webpack doesn't haev HTML entry points and I simply don't give a shit about the details of how the CSS is loaded — I just want to write my frontend and be done with it because there's some urgency in getting my current project finished.). I know that I said something similar before, but I'm really frustrated right now after this issue has taken up multiple days when I *finally* have some time to work on my project, and it's awful to just have this sort of thing appear and eat up all my time. I'm at my wit's end here, and I might need to just forget any hope of doing frontend development. I haven't felt this frustrated with programming at any point in my life that I can remember, even with older frontend systems, and I've programmed for probably 8 years now: even C++, which has a reputation for cryptic and useless errors, has never been this bad! I just want a no-fuss way of including libraries from NPM (including CSS libraries, like Bootstrap) on my frontend, and of deduplicating my code (e.g. a shared header across each HTML page).

EDIT: It looks like there are other people with similar concerns:
 - https://github.com/parcel-bundler/parcel/issues/5191 (Schrödinger's bugs)
 - https://github.com/parcel-bundler/parcel/issues/4104 (bad error messages)

**Comment by DeMoorJasper** [comment]
Clearing cache might fix it, a minimal reproduction would help us reproduce this.

This is also likely a bug in Parcel itself and not your code, otherwise you'd get a very nice error message...

We spent a lot of hours on error messages so please don't be that guy, that expects instant free support from open source software and doesn't even ask nicely

**Reply by john01dav** [reply]
> a minimal reproduction would help us reproduce this.

I fully understand how useful minimal reproductions can be — that's why making one is the second step in my debugging process. Unfortunately, due to the fact that these bugs seem to come and go with no discernable-to-me pattern, I can't do that successfully. When I take a piece of my code and put it by itself in its own project, the bug disappears. When I try to do controlled experiments to see what happens when I change parts of my code, bizarre things happen (e.g. copy/paste bad code into a working file then it suddenly works, brand new errors appear that I have no clue what to do about), and when I try to control for this it just gets messier and messier until the whole thing becomes an intractable mess of half-working code with no insight.

> Clearing cache might fix it, 

As I'm working on my code, I'm deleting node_modules, .parcel-cache, and sometimes package-lock.json with every build. It takes longer, but without doing this caching errors are on the table so I can't isolate what's happening.

> We spent a lot of hours on error messages so please don't be that guy, that expects instant free support from open source software and doesn't even ask nicely

Sorry if I seem rude — that's not the intention; I'm just feeling frustrated as it's been one thing after another.  I want to be productive, and these issues are stopping me from doing so. If I got a bug every once in awhile and had to wait for a reply while I work on something else, I could work with that, but when for every hour of actually working on my code I need to spend 12 hours debugging parcel, frankly I don't think that I can feel good about that — even if no one has done anything wrong.

**Reply by 101arrowz** [reply]
Parcel 2 isn't stable and won't be for a while. It's quite literally years in the making, and just recently a beta was released. The team is smaller, so this is to be expected. The best thing you can do to accelerate bugfixes is investigate the code yourself.

You should be able to delete just `.parcel-cache` to resolve all caching issues (I understand that there are many) rather than also deleting your lockfile and `node_modules`, since Parcel doesn't touch those. You can also use [`--no-cache`](https://v2.parceljs.org/features/cli/#general-parameters).

**Reply by aakoch** [reply]
I came across this _after_ I upgraded to v2. Which is kind of funny, but also not.

**Reply by EazzieJr** [reply]
It still doesn't work. I don't know what to do.

**Reply by IshiniKiridena** [reply]
Deleting the `.parcel-cache` worked for me... Thank u

**Comment by Meandmybadself** [comment]
I've run into this issue before & it was a result of having my application's entry point (index.js), as defined in package.json the same as another included file in the project (index.js, as included by index.html).

**Comment by kubikill** [comment]
I've run into this issue just now, it turned out it was caused by my entry HTML file that had an `<img>` tag with an empty src attribute. I fixed it by just removing the src attribute entirely.

**Reply by mulrooneydesign** [reply]
I had the same issues just too. However when I removed a tag that had an empty attribute the error resolved. This tag:

`<link rel="stylesheet" href="">`

This was in a new empty project with nothing but a stock html skeleton and a console.log() in a ./src/index.ts file.

**Comment by revilofr** [comment]
Hi, I can confirm that removing empty attributes worked for me too.
It mainly impacts users that love to use HTML boilerplates like me.

**Comment by jaredh159** [comment]
fwiw, i got this error recently too, and removing `--no-source-maps` from the cli invocation fixed it for me. ¯\\_(ツ)_/¯
